sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala

```diff
@@ -17,11 +17,14 @@
 
 package org.apache.spark.sql.execution.streaming
 
+import com.fasterxml.jackson.annotation.JsonTypeInfo
+
 /**
  * An offset is a monotonically increasing metric used to track progress in the computation of a
  * stream. An [[Offset]] must be comparable, and the result of `compareTo` must be consistent
  * with `equals` and `hashcode`.
  */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
```

Comment:
That's why Jackson has javadocs. :-) You read it once instead of having to understand how every single type ends up in json by having to read the specific manual serialization code for that type...

Anyway, to avoid copy & paste of this kind of code everywhere, it would be probably better to have a base trait for "traits that will be serialized in events and might have multiple implementations", and have that trait have this annotation (like `SparkListenerEvent` does).